File: src/algorithms/pathfinding/graph-traversal/bfs-exploration/bfs-exploration_test.py
import importlib
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "sources"))
import sys

bfs_exploration_mod = importlib.import_module("bfs-exploration")
bfs_exploration = bfs_exploration_mod.bfs_exploration


def make_empty_grid(rows, cols):
    return [[{"type": "empty"} for _ in range(cols)] for _ in range(rows)]


def set_cell(grid, row, col, cell_type):
    grid[row][col]["type"] = cell_type


def test_visits_all_cells_in_open_grid():
    grid = make_empty_grid(3, 3)
    set_cell(grid, 0, 0, "start")
    result = bfs_exploration(grid, (0, 0))
    assert len(result["visited"]) == 9


def test_starts_with_start_cell():
    grid = make_empty_grid(3, 3)
    set_cell(grid, 1, 1, "start")
    result = bfs_exploration(grid, (1, 1))
    assert result["visited"][0] == (1, 1)


def test_does_not_visit_wall_cells():
    grid = make_empty_grid(3, 3)
    set_cell(grid, 0, 1, "wall")
    set_cell(grid, 1, 0, "wall")
    set_cell(grid, 1, 1, "wall")
    result = bfs_exploration(grid, (0, 0))
    assert len(result["visited"]) == 1


def test_visits_only_reachable_cells():
    grid = make_empty_grid(4, 4)
    set_cell(grid, 0, 0, "start")
    for wall_row in range(4):
        set_cell(grid, wall_row, 2, "wall")
    result = bfs_exploration(grid, (0, 0))
    assert len(result["visited"]) == 8


def test_handles_1x1_grid():
    grid = make_empty_grid(1, 1)
    set_cell(grid, 0, 0, "start")
    result = bfs_exploration(grid, (0, 0))
    assert len(result["visited"]) == 1
    assert result["layers"] == 1


def test_no_cell_visited_twice():
    grid = make_empty_grid(4, 4)
    set_cell(grid, 0, 0, "start")
    result = bfs_exploration(grid, (0, 0))
    visited_set = set(result["visited"])
    assert len(visited_set) == len(result["visited"])


def test_visits_linear_corridor_in_order():
    grid = make_empty_grid(5, 1)
    set_cell(grid, 0, 0, "start")
    result = bfs_exploration(grid, (0, 0))
    for visit_index, cell in enumerate(result["visited"]):
        assert cell == (visit_index, 0)


if __name__ == "__main__":
    test_visits_all_cells_in_open_grid()
    test_starts_with_start_cell()
    test_does_not_visit_wall_cells()
    test_visits_only_reachable_cells()
    test_handles_1x1_grid()
    test_no_cell_visited_twice()
    test_visits_linear_corridor_in_order()
    print("All tests passed!")

File: src/algorithms/pathfinding/graph-traversal/bfs-exploration/__tests__/BfsExploration_test.java
import java.util.*;

// javac BfsExploration.java BfsExploration_test.java && java -ea BfsExploration_test
public class BfsExploration_test {

    static int[][] makeEmptyGrid(int rows, int cols) {
        return new int[rows][cols];
    }

    static void setWall(int[][] grid, int row, int col) {
        grid[row][col] = 1;
    }

    @SuppressWarnings("unchecked")
    public static void main(String[] args) {
        // Test: visits all cells in open grid
        {
            int[][] grid = makeEmptyGrid(3, 3);
            Map<String, Object> result = BfsExploration.bfsExploration(grid, new int[]{0, 0});
            List<?> visited = (List<?>) result.get("visited");
            assert visited.size() == 9 : "Expected 9 visited cells, got " + visited.size();
        }

        // Test: starts with start cell
        {
            int[][] grid = makeEmptyGrid(3, 3);
            Map<String, Object> result = BfsExploration.bfsExploration(grid, new int[]{1, 1});
            List<int[]> visited = (List<int[]>) result.get("visited");
            assert visited.get(0)[0] == 1 && visited.get(0)[1] == 1
                : "Expected first visited to be [1,1]";
        }

        // Test: does not visit wall cells
        {
            int[][] grid = makeEmptyGrid(3, 3);
            setWall(grid, 0, 1);
            setWall(grid, 1, 0);
            setWall(grid, 1, 1);
            Map<String, Object> result = BfsExploration.bfsExploration(grid, new int[]{0, 0});
            List<?> visited = (List<?>) result.get("visited");
            assert visited.size() == 1 : "Expected 1 visited cell, got " + visited.size();
        }

        // Test: visits only reachable cells
        {
            int[][] grid = makeEmptyGrid(4, 4);
            for (int wallRow = 0; wallRow < 4; wallRow++) setWall(grid, wallRow, 2);
            Map<String, Object> result = BfsExploration.bfsExploration(grid, new int[]{0, 0});
            List<?> visited = (List<?>) result.get("visited");
            assert visited.size() == 8 : "Expected 8 visited cells, got " + visited.size();
        }

        // Test: handles 1x1 grid
        {
            int[][] grid = makeEmptyGrid(1, 1);
            Map<String, Object> result = BfsExploration.bfsExploration(grid, new int[]{0, 0});
            List<?> visited = (List<?>) result.get("visited");
            assert visited.size() == 1 : "Expected 1 visited cell";
        }

        System.out.println("All tests passed!");
    }
}
